sys/pool.c

```diff
@@ -0,0 +1,203 @@
+#include <stdc.h>
+#include <vm.h>
+#include <physmem.h>
+#include <queue.h>
+#include <bitstring.h>
+#include <common.h>
+#include <pool.h>
+
+#define PI_MAGIC_WORD 0xcafebabe
+#define PP_FREED_WORD 0xdeadbeef
+
+typedef struct pool_slab {
+  LIST_ENTRY(pool_slab) ph_slablist; /* pool slab list */
+  vm_page_t *ph_page;                /* page containing this slab */
+  uint16_t ph_nused;                 /* # of items in use */
+  uint16_t ph_nfree;  /* # of free (and available) items, a bit redundant but
+                         there would be padding anyway */
+  uint16_t ph_ntotal; /* total number of chunks*/
+  uint16_t ph_start;  /* start offset in page */
+  bitstr_t ph_bitmap[0];
+} pool_slab_t;
+
+typedef struct pool_item {
+  uint32_t pi_canary;   /* PI_MAGIC_WORD by default, if modified then memory
+                           corruption has occured */
+  pool_slab_t *pi_slab; /* pointer to slab containing this item */
+  unsigned long pi_data[0];
+} pool_item_t;
+
+static pool_item_t *get_pi_at_idx(pool_slab_t *slab, unsigned i, size_t size) {
+  return (pool_item_t *)((intptr_t)slab + slab->ph_start +
+                         i * (size + sizeof(pool_item_t)));
+}
+
+static pool_slab_t *create_slab(pool_t *pool) {
+  /* log("create_slab: pool = %p, pp_itemsize = %d", pool, pool->pp_itemsize);
+   */
+  vm_page_t *page_for_slab = pm_alloc(1);
+  pool_slab_t *slab = (pool_slab_t *)page_for_slab->vaddr;
+  slab->ph_page = page_for_slab;
+  slab->ph_nused = 0;
+  /* sizeof(pool_slab_t) + n * (sizeof(pool_item_t) + size)
+   * + ((n + 31) / 32) * 4 <= PAGESIZE */
+  slab->ph_ntotal = (8 * (PAGESIZE - sizeof(pool_slab_t)) - 31) /
+                    (8 * (sizeof(pool_item_t) + pool->pp_itemsize) + 1);
+  slab->ph_nfree = slab->ph_ntotal;
+  slab->ph_start = sizeof(pool_slab_t) + howmany(slab->ph_ntotal, 32) * 4;
+  memset(slab->ph_bitmap, 0, bitstr_size(slab->ph_ntotal));
+  for (int i = 0; i < slab->ph_ntotal; i++) {
+    pool_item_t *curr_pi = get_pi_at_idx(slab, i, pool->pp_itemsize);
+    /* log("Still alive, are we? %d", i); */
+    curr_pi->pi_slab = slab;
+    curr_pi->pi_canary = PI_MAGIC_WORD;
+    if (pool->pp_ctor)
+      pool->pp_ctor(curr_pi->pi_data, pool->pp_itemsize);
+  }
+  return slab;
+}
+
+static void destroy_slab(pool_t *pool, pool_slab_t *slab) {
+  log("destroy_slab: pool = %p, slab = %p", pool, slab);
+  for (int i = 0; i < slab->ph_ntotal; i++) {
+    pool_item_t *curr_pi = get_pi_at_idx(slab, i, pool->pp_itemsize);
+    if (pool->pp_dtor)
+      pool->pp_dtor(curr_pi->pi_data, pool->pp_itemsize);
```

Comment:
Is it necessary to pass `pool->pp_itemsize` here and in other places? What if it the value was present in `pool_slab` structure as well?